wpinet/src/main/native/cpp/WebSocketSerializer.h

```diff
@@ -0,0 +1,270 @@
+    MakeReq&& makeReq,
+    std::function<void(std::span<uv::Buffer>, uv::Error)> callback) {
+  auto frameIt = frames.begin();
+  auto frameEnd = frames.end();
+  while (frameIt != frameEnd) {
+    auto frameStart = frameIt;
+
+    // build buffers to send
+    SerializedFrames sendFrames;
+    SmallVector<int, 32> frameOffs;
+    int numBytes = 0;
+    while (frameIt != frameEnd) {
+      frameOffs.emplace_back(numBytes);
+      numBytes += sendFrames.AddFrame(*frameIt++, server);
+      if ((server && (numBytes >= 65536 || frameOffs.size() > 32)) ||
+          (!server && numBytes >= 8192)) {
+        // don't waste too much memory or effort on header generation or masking
+        break;
+      }
+    }
+
+    // try to send
+    int sentBytes = stream.TryWrite(sendFrames.m_bufs);
+
+    if (sentBytes == 0) {
+      // we haven't started a frame yet; clean up any bufs that have actually
+      // sent, and return unsent frames
+      SmallVector<uv::Buffer, 4> bufs;
+      for (auto it = frames.begin(); it != frameStart; ++it) {
+        bufs.append(it->data.begin(), it->data.end());
+      }
+      callback(bufs, {});
+      return {frameStart, frameEnd};
+    } else if (sentBytes < 0) {
+      // error
+      SmallVector<uv::Buffer, 4> bufs;
+      for (auto&& frame : frames) {
+        bufs.append(frame.data.begin(), frame.data.end());
+      }
+      callback(bufs, uv::Error{sentBytes});
+      return {};
+    } else if (sentBytes != numBytes) {
+      // we didn't send everything; deal with the leftovers
+
+      // figure out what the last (partially) frame sent actually was
+      auto offIt = frameOffs.begin();
+      auto offEnd = frameOffs.end();
+      bool isFin = true;
+      while (offIt != offEnd && *offIt < sentBytes) {
+        ++offIt;
+        isFin = (frameStart->opcode & WebSocket::kFlagFin) != 0;
+        ++frameStart;
+      }
+
+      if (offIt != offEnd && *offIt == sentBytes && isFin) {
+        // we finished at a normal FIN frame boundary; no need for a Write()
+        SmallVector<uv::Buffer, 4> bufs;
+        for (auto it = frames.begin(); it != frameStart; ++it) {
+          bufs.append(it->data.begin(), it->data.end());
+        }
+        callback(bufs, {});
+        return {frameStart, frameEnd};
+      }
+
+      // build a list of buffers to send as a normal write:
+      SmallVector<uv::Buffer, 4> writeBufs;
+      auto bufIt = sendFrames.m_bufs.begin();
+      auto bufEnd = sendFrames.m_bufs.end();
+
+      // start with the remaining portion of the last buffer actually sent
+      int pos = 0;
+      while (bufIt != bufEnd && pos < sentBytes) {
+        pos += (bufIt++)->len;
+      }
+      if (bufIt != sendFrames.m_bufs.begin() && pos != sentBytes) {
+        writeBufs.emplace_back(
+            wpi::take_back((bufIt - 1)->bytes(), pos - sentBytes));
+      }
+
+      // continue through the last buffer of the last partial frame
+      while (bufIt != bufEnd && offIt != offEnd && pos < *offIt) {
```

Comment:
Minor: Since `offIt` isn't changed by the loop, could the check to `offIt != offEnd` be moved to an if-statement around the loop (and potentially combined with the check immediately after)?